### PR TITLE
adds first frame handling timeout

### LINKED ISSUE #1025: Limit max time for initial SETUP or RESUME frame to be received
The spec says:

> Immediately upon successful connection, the client MUST send either a SETUP or RESUME frame with Stream ID of 0. Any other frame received that is NOT a SETUP|RESUME frame or a SETUP|RESUME frame with a Stream ID > 0, MUST cause the server to send an ERROR[INVALID_SETUP] and close the connection.

The server should have a configurable limit that determines how long to wait for the SETUP or RESUME frame to materialize or close the connection otherwise. This timeout should probably have some default value that is enforced out of the box but can be customized.

**Note:** originally reported in the Spring Framework issue tracker, as https://github.com/spring-projects/spring-framework/issues/27373, along with a [repro project](https://github.com/mostroverkhov/springboot-repros) (see `/dos_idle_connections` and the `server.sh` script).

### PR DESCRIPTION
**Pull request by OlegDokuka**
fixes #1025 

Signed-off-by: Oleh Dokuka <oleh.dokuka@icloud.com>
Signed-off-by: Oleh Dokuka <odokuka@vmware.com>

### CODE REVIEW

File: rsocket-core/src/main/java/io/rsocket/core/RSocketServer.java [hunk 1]
```diff
@@ -223,6 +225,22 @@ public RSocketServer maxInboundPayloadSize(int maxInboundPayloadSize) {
     return this;
   }
 
+  /**
+   * Specifies timeout for the first incoming frame from the accepted connection.
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/core/RSocketServer.java:229]:
```suggestion
   * Specify the max time to wait for the first frame (e.g. {@code SETUP}) on an accepted connection.
```

File: rsocket-core/src/main/java/io/rsocket/core/RSocketServer.java [hunk 2]
```diff
@@ -223,6 +225,22 @@ public RSocketServer maxInboundPayloadSize(int maxInboundPayloadSize) {
...
+   *
+   * <p>By default this is set to 1 minute.
+   *
+   * @param timeout duration
+   * @return the same instance for method chaining
+   */
+  public RSocketServer setupHandlingTimeout(Duration timeout) {
```
Comment (rstoyanchev) [inline rsocket-core/src/main/java/io/rsocket/core/RSocketServer.java:236]: How about calling this `maxTimeToFirstFrame`?

**Review by rstoyanchev — COMMENTED** [review]
Looks good. A couple of minor comments.

File: rsocket-core/src/main/java/io/rsocket/core/RSocketServer.java (hunk 2 above)
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/core/RSocketServer.java:236]: Sound good to me!

Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/core/RSocketServer.java:236]:
```suggestion
  public RSocketServer maxTimeToFirstFrame(Duration timeout) {
```

File: rsocket-core/src/test/java/io/rsocket/core/RSocketServerTest.java [hunk 3]
```diff
@@ -60,6 +62,32 @@ public void unexpectedFramesBeforeSetupFrame() {
         .hasNoLeaks();
   }
 
+  @Test
+  public void timeoutOnNoFirstFrame() {
+    final VirtualTimeScheduler scheduler = VirtualTimeScheduler.getOrSet();
+    try {
+      TestServerTransport transport = new TestServerTransport();
+      RSocketServer.create().setupHandlingTimeout(Duration.ofMinutes(2)).bind(transport).block();
```
Comment (OlegDokuka) [inline rsocket-core/src/test/java/io/rsocket/core/RSocketServerTest.java:70]:
```suggestion
      RSocketServer.create().maxTimeToFirstFrame(Duration.ofMinutes(2)).bind(transport).block();
```

**Pull request merged**